// react-rainbow-components/esm/components/Lookup/selectedValue.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import copy from 'clipboard-copy';
import formatValue from './helpers/formatValue';
import RenderIf from '../RenderIf/index';
import CloseIcon from '../Chip/closeIcon';
import ButtonIcon from '../ButtonIcon/index';
import StyledReadOnlySelectedInput from './styled/input';
import StyledSelectedInput from './styled/selectedInput';
import StyledSelectedValueIcon from './styled/selectedValueIcon';
import StyledSelectedValueContainer from './styled/selectedValueContainer';
import StyledSelectedValueClearButton from './styled/selectedValueClearButton';

var SelectedValue = function (_Component) {
  _inherits(SelectedValue, _Component);

  var _super = _createSuper(SelectedValue);

  function SelectedValue(props) {
    var _this;

    _classCallCheck(this, SelectedValue);

    _this = _super.call(this, props);
    _this.inputRef = React.createRef();
    _this.handleFocus = _this.handleFocus.bind(_assertThisInitialized(_this));
    return _this;
  }

  _createClass(SelectedValue, [{
    key: "handleFocus",
    value: function handleFocus() {
      var value = this.props.value;

      var _formatValue = formatValue(value),
          label = _formatValue.label;

      copy(label);
    }
  }, {
    key: "focus",
    value: function focus() {
      this.inputRef.current.focus();
    }
  }, {
    key: "click",
    value: function click() {
      this.inputRef.current.click();
    }
  }, {
    key: "blur",
    value: function blur() {
      this.inputRef.current.blur();
    }
  }, {
    key: "renderInput",
    value: function renderInput() {
      var _this$props = this.props,
          id = _this$props.id,
          name = _this$props.name,
          value = _this$props.value,
          disabled = _this$props.disabled,
          tabIndex = _this$props.tabIndex,
          readOnly = _this$props.readOnly,
          required = _this$props.required,
          onClick = _this$props.onClick,
          errorMessageId = _this$props.errorMessageId,
          error = _this$props.error;

      var _formatValue2 = formatValue(value),
          label = _formatValue2.label,
          icon = _formatValue2.icon;

      if (readOnly) {
        return React.createElement(StyledReadOnlySelectedInput, {
          id: id,
          name: name,
          type: "text",
          value: label,
          tabIndex: tabIndex,
          onFocus: this.handleFocus,
          onClick: onClick,
          disabled: disabled,
          readOnly: true,
          "aria-describedby": errorMessageId,
          required: required,
          ref: this.inputRef,
          iconPosition: "left",
          icon: icon,
          error: error
        });
      }

      return React.createElement(StyledSelectedInput, {
        id: id,
        name: name,
        type: "text",
        value: label,
        tabIndex: tabIndex,
        onFocus: this.handleFocus,
        onClick: onClick,
        disabled: disabled,
        readOnly: true,
        "aria-describedby": errorMessageId,
        required: required,
        ref: this.inputRef,
        iconPosition: "left",
        icon: icon,
        error: error
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          value = _this$props2.value,
          disabled = _this$props2.disabled,
          readOnly = _this$props2.readOnly,
          onClearValue = _this$props2.onClearValue;

      var _formatValue3 = formatValue(value),
          icon = _formatValue3.icon;

      return React.createElement(StyledSelectedValueContainer, {
        readOnly: readOnly
      }, React.createElement(RenderIf, {
        isTrue: icon
      }, React.createElement(StyledSelectedValueIcon, {
        readOnly: readOnly
      }, icon)), this.renderInput(), React.createElement(RenderIf, {
        isTrue: !(readOnly || disabled)
      }, React.createElement(StyledSelectedValueClearButton, null, React.createElement(ButtonIcon, {
        assistiveText: "clear",
        size: "small",
        title: "Remove selected option",
        icon: React.createElement(CloseIcon, {
          color: "#576574"
        }),
        onClick: onClearValue
      }))));
    }
  }]);

  return SelectedValue;
}(Component);

export { SelectedValue as default };
SelectedValue.propTypes = {
  value: PropTypes.oneOfType([PropTypes.shape({
    label: PropTypes.string,
    description: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
    icon: PropTypes.node
  }), PropTypes.string]),
  id: PropTypes.string,
  name: PropTypes.string,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  readOnly: PropTypes.bool,
  tabIndex: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  onClick: PropTypes.func,
  onClearValue: PropTypes.func,
  errorMessageId: PropTypes.string,
  error: PropTypes.oneOfType([PropTypes.string, PropTypes.node])
};
SelectedValue.defaultProps = {
  value: undefined,
  name: undefined,
  id: undefined,
  disabled: false,
  required: false,
  readOnly: false,
  tabIndex: undefined,
  onClick: function onClick() {},
  onClearValue: undefined,
  errorMessageId: undefined,
  error: undefined
};
